File: src/main/java/io/github/komdosh/alarms/services/impl/AlarmServiceImpl.java
package io.github.komdosh.alarms.services.impl;

import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.Messages;
import io.github.komdosh.alarms.services.AlarmService;

import javax.swing.*;
import java.lang.reflect.InvocationTargetException;
import java.util.concurrent.TimeUnit;
import java.util.logging.Logger;

public class AlarmServiceImpl implements AlarmService {
    private boolean isEnabled = false;
    private Thread executionTimeoutLoop;
    private Logger logger = Logger.getLogger(AlarmService.class.getName());

    public void setAlertInSeconds(Project p, long delay, String dimension) {
        long delayNumberForText;
        if (dimension.equalsIgnoreCase("minutes")) {
            delayNumberForText = TimeUnit.SECONDS.toMinutes(delay);
        } else {
            delayNumberForText = delay;
        }

        Messages.showMessageDialog(p, "Alarm schedule for every " + delayNumberForText + " " + dimension,
            "Alarm Is Set", Messages.getInformationIcon());

        isEnabled = true;

        setTimeout(() -> Messages.showMessageDialog(p, "You should take a break," +
                " alarm is set for every " + delay + " " + dimension,
            "Time to Relax", Messages.getInformationIcon()), delay);
    }

    private void setTimeout(Runnable s, long delay) {
        if (!isEnabled) {
            return;
        }

        executionTimeoutLoop = new Thread(() -> {
            setExecution(s, delay);
        });

        executionTimeoutLoop.start();
    }

    private void setExecution(Runnable s, long delay) {
        try {
            Thread.sleep(TimeUnit.SECONDS.toMillis(delay));
            SwingUtilities.invokeAndWait(s);
            setTimeout(s, delay);
        } catch (InterruptedException e) {
            logger.info("io.github.komdosh.alarms: Alarm Thread Interrupted");
        } catch (InvocationTargetException e) {
            logger.info("io.github.komdosh.alarms: InvocationTargetException");
        }
    }

    public void disableAlerts() {
        isEnabled = false;
        executionTimeoutLoop.interrupt();
    }
}

File: src/main/kotlin/io/github/komdosh/alarms/services/impl/AlarmServiceImpl.kt
package io.github.komdosh.alarms.services.impl

import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.application.ModalityState
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.Messages
import io.github.komdosh.alarms.services.AlarmService
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit


class AlarmServiceImpl : AlarmService {
    private var scheduler = Executors.newSingleThreadScheduledExecutor()

    override fun setAlertInSeconds(p: Project, timeUnit: TimeUnit, value: Long) {
        val textDimension = when (timeUnit) {
            TimeUnit.SECONDS -> "seconds"
            TimeUnit.MINUTES -> "minutes"
            else -> "unknown time unit"
        }

        val delay: Long = if (timeUnit == TimeUnit.SECONDS) {
            TimeUnit.SECONDS.toMinutes(value)
        } else {
            value
        }

        Messages.showMessageDialog(p, "Alarm schedule for every $value $textDimension",
                "Alarm is Set", Messages.getInformationIcon())

        scheduler.scheduleWithFixedDelay({
            ApplicationManager.getApplication().invokeAndWait({
                Messages.showMessageDialog(p, "You should take a break, alarm is set for every $value $textDimension",
                        "Time to Relax", Messages.getInformationIcon())
            }, ModalityState.NON_MODAL)
        }, delay, delay, TimeUnit.SECONDS)
    }

    override fun disableAlerts(p: Project) {
        Messages.showMessageDialog(p, "Alarm is disabled",
                "Alarm is Disabled", Messages.getInformationIcon())
        scheduler.shutdown()
        scheduler = Executors.newSingleThreadScheduledExecutor()
    }
}
